File: 1334/solution.ts
function findTheCity(
  n: number,
  edges: number[][],
  distanceThreshold: number
): number {
  const dist: number[][] = Array.from({ length: n }, () =>
    Array(n).fill(Number.POSITIVE_INFINITY)
  );

  for (let i = 0; i < n; i++) {
    dist[i][i] = 0;
  }

  for (let [x, y, d] of edges) {
    dist[x][y] = d;
    dist[y][x] = d;
  }

  for (let k = 0; k < n; k++) {
    for (let i = 0; i < n; i++) {
      for (let j = 0; j < n; j++) {
        if (dist[i][k] + dist[k][j] < dist[i][j]) {
          dist[i][j] = dist[i][k] + dist[k][j];
        }
      }
    }
  }

  let res = [0, n];
  for (let i = 0; i < n; i++) {
    let reachableCities = 0;
    for (let j = 0; j < n; j++) {
      if (i !== j && dist[i][j] <= distanceThreshold) {
        reachableCities++;
      }
    }
    if (reachableCities <= res[1]) {
      res = [i, reachableCities];
    }
  }

  return res[0];
}

File: 1334/solution.py
from collections import defaultdict
from heapq import heappop, heappush


class Solution:
    def findTheCity(
        self, n: int, edges: list[list[int]], distanceThreshold: int
    ) -> int:
        res = [1, float("inf")]
        graph = defaultdict(list)
        for x, y, dist in edges:
            graph[x].append([y, dist])
            graph[y].append([x, dist])
        for i in range(n):
            visited = set()
            heap = [[0, i]]
            while heap:
                dist, city = heappop(heap)
                if city in visited:
                    continue
                visited.add(city)
                for child, d in graph[city]:
                    if child not in visited and dist + d <= distanceThreshold:
                        heappush(heap, [dist + d, child])
            if len(visited) - 1 <= res[1]:
                res = [i, len(visited) - 1]
        return res[0]
